Sources/MongoSwift/MongoClient.swift

```diff
@@ -252,41 +274,109 @@ public class MongoClient {
+     * - `LogicError` if `syncShutdown` or `shutdown` has already been called on this client.
+     */
+    public func shutdown(queue: DispatchQueue = .global(), _ callback: @escaping (Error?) -> Void) {
+        let error: Error? = self.stateLock.withLock {
+            guard self.state == .upAndRunning else {
+                return Self.ShutdownClientError
+            }
+            self.state = .shuttingDown
+            return nil
+        }
+
+        guard error == nil else {
+            return callback(error)
+        }
+
+        // close connection pool
         self.operationExecutor.execute {
             self.connectionPool.shutdown()
-            self.isClosed = true
-        }
-        .flatMap {
-            self.operationExecutor.close()
+            // ignore errors; this shouldn't fail unless the user already shut down the ELG
+        }.whenComplete { _ in
+            self.operationExecutor.shutdown(queue: queue) { error in
+                self.stateLock.withLock {
+                    self.state = .shutDown
+                }
+                callback(error)
+            }
         }
     }
 
     /**
-     * Shuts this `MongoClient` down in a blocking fashion, closing all connections to the server and cleaning up
-     * internal state.
+     * Shuts this `MongoClient` down in a blocking fashion, closing all connections to the server and shutting down the
+     * underlying thread pool.
+     *
+     * If `.createNew` was specified as the `EventLoopGroupProvider`, this method will also shut down the underlying
+     * `EventLoopGroup`. If `.shared` was specified, the provided `EventLoopGroup` must not be shut down until after
+     * the client has finished shutting down.
      *
      * Call this method exactly once when you are finished using the client. You must ensure that all operations
      * using the client have completed before calling this.
      *
-     * This method must complete before the `EventLoopGroup` provided to this client's constructor is shut down.
+     * Since this method is blocking, you must not call it on an `EventLoop`.
+     *
+     * - Throws:
+     *   - `LogicError` if `syncShutdown` or `shutdown` has already been called on this client.
      */
-    public func syncShutdown() {
-        self.connectionPool.shutdown()
-        self.isClosed = true
-        // TODO: SWIFT-349 log any errors encountered here.
-        try? self.operationExecutor.syncClose()
+    public func syncShutdown() throws {
+        if let eventLoop = MultiThreadedEventLoopGroup.currentEventLoop {
+            preconditionFailure("""
+            BUG DETECTED: syncShutdown() must not be called when on an EventLoop.
+            Calling syncShutdown() on any EventLoop can lead to deadlocks.
+            Current eventLoop: \(eventLoop)
+            """)
+        }
+        let errorStorageLock = Lock()
+        var errorStorage: Error?
+        let continuation = DispatchWorkItem {}
```

Comment:
This `DispatchWorkItem` seems like it could have been useful for the cursor stuff. The way it's being used here is very similar to a channel which I remember we wanted. Looking at the cursor code now, I can't really remember why exactly I thought a channel was necessary though.